src/compiler/program.ts

```diff
@@ -99,20 +106,27 @@ namespace ts {
         const directoryExists = !onlyRecordFailures && directoryProbablyExists(candidate, host);
         if (directoryExists && host.fileExists(packageJsonPath)) {
 
-            let jsonContent: { typings?: string };
+            let jsonContent: { typings?: string; main?: string };
 
             try {
                 const jsonText = host.readFile(packageJsonPath);
-                jsonContent = jsonText ? <{ typings?: string }>JSON.parse(jsonText) : { typings: undefined };
+                jsonContent = jsonText ? <{ typings?: string; main?: string }>JSON.parse(jsonText) : { typings: undefined };
             }
             catch (e) {
-                // gracefully handle if readFile fails or returns not JSON 
+                // gracefully handle if readFile fails or returns not JSON
                 jsonContent = { typings: undefined };
             }
 
             if (typeof jsonContent.typings === "string") {
                 const path = normalizePath(combinePaths(candidate, jsonContent.typings));
-                const result = loadNodeModuleFromFile(extensions, path, failedLookupLocation, !directoryProbablyExists(getDirectoryPath(path), host), host);
+                const result = loadNodeModuleFromFile(/*don't add extension*/[""], path, failedLookupLocation, !directoryProbablyExists(getDirectoryPath(path), host), host);
+                if (result) {
+                    return result;
+                }
+            }
+            if (typeof jsonContent.main === "string") {

```

Comment:
`else if` ? do you want it to fall back to js and result in an error of unsupported file extension or just can not resolve module? i think we should not pick the js file if there is a .ts one.